src/Umbraco.Core/Models/KeyValue.cs

```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.Runtime.Serialization;
+using Umbraco.Core.Models.Entities;
+
+namespace Umbraco.Core.Models
+{
+    /// <summary>
+    /// Implements <see cref="IKeyValue"/>.
+    /// </summary>
+    [Serializable]
+    [DataContract(IsReference = true)]
+    public class KeyValue : EntityBase, IKeyValue
```

Comment:
This should explicitly implement `IEntity` and then explicitly implement `IEntity.HasIdentity` to return it's own logic of whether this entity has an identity. (explicit implementation is the same as overriding a base class). Then we don't have to make the `Save` method in RepositoryBase virtual and we don't have to override that method in the KeyValueRepository.

Reply:
Had some trouble with this one.  The problem being that the key value "entity" doesn't actually have an identity like an integer (that will be zero if the record doesn't already exist) - there's no "id" primary key in `umbracoKeyValue`, the key itself is the primary key.

So when saving an instance of a key value pair, the only way to see if it's a new or updated entity, is to attempt to retrieve it by key and see if it exists.  If so, it's an update, if not an insert.

And that retrieval needs to be in the repository values - the `KeyValue` model class doesn't know from it's own properties if it's a new or updated value, and hence can't implement `HasIdentity`.

So have left this as is for now, pending any further thoughts or feedback...

Reply:
Ah yes i see what you mean because even if its not persisted you of course have to assign a key and we cannot just know if the entity exists if that key is null or not, makes sense! Oh well, would have been nice :) we'll keep what you have done